test/k8s: keep configmap across upgrade test

aanm opened the pull request:
test/k8s: keep configmap across upgrade test

Since we have new options set for new Cilium installations those
might break the connectivity when doing upgrades.

To ideally test an upgrade, we should keep the existing ConfigMap
from the previous the version.

This also fixes the upgrade guide using helm as using `--set
config.enabled=false` deploys Cilium without its configuration map,
deleting the existing one from the cluster.

Also fixes the Cilium DaemonSet since its upgrade does not work using
kubectl apply given that we have introduced a new liveness and readiness
probes in the DaemonSet. Added upgrade notes to keep the old probes
while users perform the upgrades.

Fixes: d613dea5d524 ("install/kubernetes: use HTTP for agent {liveness,readiness}Probe")
Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
test/k8s: keep configmap across upgrade test

Since we have new options set for new Cilium installations those
might break the connectivity when doing upgrades.

To ideally test an upgrade, we should keep the existing ConfigMap
from the previous the version.

This also fixes the upgrade guide using helm as using `--set
config.enabled=false` deploys Cilium without its configuration map,
deleting the existing one from the cluster.

Also fixes the Cilium DaemonSet since its upgrade does not work using
kubectl apply given that we have introduced a new liveness and readiness
probes in the DaemonSet. Added upgrade notes to keep the old probes
while users perform the upgrades.

Fixes: d613dea5d524 ("install/kubernetes: use HTTP for agent {liveness,readiness}Probe")
Signed-off-by: André Martins <andre@cilium.io>

joestringer reviewed (approved):
There's another option which is `--reuse-values` which I think we should start encouraging more widely including in the upgrade docs.

But this actually matches our upgrade docs so :ok_hand:

joestringer reviewed (approved):
I think this is good enough to fix the upgrade for v1.7 and v1.8 branches, then during v1.9 cycle we can take a long hard look at our Helm charts & upgrade instructions to figure out the long-term solution here.

joestringer commented on Documentation/install/upgrade.rst:
```diff
@@ -209,21 +211,16 @@ configuration options for each minor version.
    ``values.yaml`` because all options are fed into the DaemonSets and
    Deployments using the `ConfigMap` which is not generated if
-   ``config.enabled=false`` is set. The above command *only* generates the
-   DaemonSet, Deployment and RBAC definitions.
+   ``config.enabled=false`` or ``config.keepCurrent=true`` are set. The above
+   command *only* generates the DaemonSet, Deployment and RBAC definitions.
```
Actually we need to drop this hunk.

aanm committed:
operator: Fix IPAM default flag

Unless the user explicitly specifies the IPAM in use, fall back to the
default cilium-operator image and use the legacy host-scope IPAM to
ensure that during upgrade from a previous release that Cilium is not
automatically migrated to the new IPAM model; that should be an explicit
opt-in step that the user performs subsequently.

Signed-off-by: André Martins <andre@cilium.io>
Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer commented on Documentation/install/upgrade.rst:
```diff
@@ -168,7 +168,9 @@ Kubernetes resources are updated accordingly to version you are upgrading to:
-    Deploy Cilium release via Helm:
+    Deploy Cilium release via Helm (Running this command will overwrite the
+    existing cluster's `ConfigMap` which might not be ideal if you want to keep
+    your existing `ConfigMap`.)
```
For micro releases, we should just use `--reuse-values`.

joestringer replied:
oh, actually AFAIU since we don't specify any `--set`, `--reuse-values` is the default.

joestringer reviewed (approved):
Minor comment that doesn't need to block merging / backporting / -rc3.

tgraf commented on Documentation/install/upgrade.rst:
```diff
@@ -168,7 +168,9 @@ Kubernetes resources are updated accordingly to version you are upgrading to:
-    Deploy Cilium release via Helm:
+    Deploy Cilium release via Helm (Running this command will overwrite the
+    existing cluster's `ConfigMap` which might not be ideal if you want to keep
+    your existing `ConfigMap`.)
```
Do we need an overall warning that this will possibly break compatibility as we will use the new helm defaults for upgrades? So far we have been assuming that the ConfigMap values would remain unchanged. The above `helm template` will overwrite it as far as I understand. Our backwards compatibility guarantee has been around not changing the defaults in the binary but default breaking changes to the new value in the ConfigMap so new deployments get the good new values without breaking existing clusters?

aanm replied:
>Do we need an overall warning that this will possibly break compatibility as we will use the new helm defaults for upgrades?

yes. I think we need to take an overall look at this section since we use `helm upgrade` in a couple places.

>So far we have been assuming that the ConfigMap values would remain unchanged. The above helm template will overwrite it as far as I understand

good point, I've added the warning in the notes section.

>Our backwards compatibility guarantee has been around not changing the defaults in the binary but default breaking changes to the new value in the ConfigMap so new deployments get the good new values without breaking existing clusters?

If I understand you correctly, yes that is supposed to be our guarantee.

tgraf reviewed (approved):
Fix for probe looks great. I have a question remaining for option A in the upgrade path.
